tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNet/AspNetWebApi2Integration.cs

```diff
@@ -96,45 +96,58 @@ internal static void UpdateSpan(IHttpControllerContext controllerContext, Span s
                 {
                 }
 
+                IDictionary<string, object> routeValues = null;
+                try
+                {
+                    routeValues = controllerContext.RouteData.Values;
+                }
+                catch
+                {
+                }
+
                 string resourceName;
 
-                if (route != null)
+                string controller = string.Empty;
+                string action = string.Empty;
+                string area = string.Empty;
+
+                if (route != null && routeValues is not null)
                 {
-                    resourceName = $"{method} {(newResourceNamesEnabled ? "/" : string.Empty)}{route.ToLowerInvariant()}";
+                    resourceName = AspNetResourceNameHelper.CalculateResourceName(
+                        httpMethod: method,
+                        routeTemplate: route,
+                        routeValues,
+                        defaults: null,
+                        out area,
+                        out controller,
+                        out action,
+                        addSlashPrefix: newResourceNamesEnabled,
+                        expandRouteTemplates: newResourceNamesEnabled && Tracer.Instance.Settings.ExpandRouteTemplatesEnabled);
                 }
                 else if (requestUri != null)
                 {
                     var cleanUri = UriHelpers.GetCleanUriPath(requestUri);
-                    resourceName = $"{method} {cleanUri.ToLowerInvariant()}";
+                    resourceName = $"{method} {cleanUri}";
                 }
                 else
                 {
                     resourceName = $"{method}";
                 }
 
-                string controller = string.Empty;
-                string action = string.Empty;
-                string area = string.Empty;
-                try
+                if (route is null && routeValues is not null)
                 {
-                    var routeValues = controllerContext.RouteData.Values;
-                    if (routeValues != null)
+                    // we weren't able to get the route template (somehow) but _were_ able to
+                    // get the route values. Not sure how this is possible, but is preexisting behaviour
```

Comment:
Can we emit telemetry for events like these to see if it ever actually happens in the wild?

Reply:
Yes, I have a task to add `generate-metrics` to the telemetry, which we could use for cases like this. Will we _ever_ be confident enough to remove it though? 🤔 😉